doc/decisions/0_drafts/keyset_types.md

```diff
@@ -0,0 +1,79 @@
+# Types of `KeySet`s
+
+## Problem
+
+A `KeySet` can be used in different ways.
+Among these are at least:
+
+- As originally intended, a set of configuration data
+- Metadata of a `Key`
+- General associative array data structure
+
+These types have slightly different requirements, and therefore may need to be treated differently.
+
+## Constraints
+
+- API duplication just to separate types should be avoided
+- API must be safe to use for each of the different types
+
+## Assumptions
+
+- The types are not different enough to warrant the overhead of entirely different APIs and data structures.
+
+## Considered Alternatives
+
+### No restrictions
+
+Leave `KeySet` entirely unrestricted, so it can be used for each use case.
+
+Sometimes there can be odd edge cases.
+For example, what should a storage plugin do, if it receives a `KeySet` without both configuration data and metadata.
+
+### Restriction based on first `Key`
+
+To effectively create different kinds of `KeySet`s, we restrict which `Key`s can be inserted into a `KeySet` based on the first `Key` in a `KeySet`.
+
+Any `Key` can be inserted into an empty `KeySet`.
+Only `Key`s matching the type of a `KeySet` can be inserted into a non-empty `KeySet`.
+The type of a `KeySet` is determined by the first `Key`, i.e. `ksAtCursor(ks, 0)`.
+Transitively, this means the type of a `KeySet` is fixed when the first `Key` is inserted and when a `KeySet` is cleared, it becomes "type-less" again.
+
+The different types of `KeySet` will be:
+
+1. Metadata:
+   Can only contain `Key`s with namespace `KEY_NS_META`.
+2. General data:
+   Can only contain `Key`s with namespace `KEY_NS_DATA`.
```

Comment:
KEY_NS_DATA should be explained somewhere, probably best in its use case.

Reply:
I think I'll add a new decision to this PR that introduces `KEY_NS_DATA`. It should be a very short and simple decision and under the constraint of this decision there is only one alternative: "Don't use `KeySet` for general data".